### ISSUE TITLE
Final Fix: `train_all_models.py`'yi Özellik Maskesi ve `config.example.yaml` Ayarlarını Kullanacak Şekilde Onar

### ISSUE THREAD

**Issue by SefaGH** [SefaGH/bearish-alpha-bot#389]
### 🎯 Amaç

Bu görev, `train-models.yml` workflow'unun arka planda sessizce başarısız olmasına neden olan kritik hataları gidermeyi hedefler. Ana eğitim script'i (`train_all_models.py`), **1)** repository'ye commit'lenmiş olan özellik seçim maskesini (`.npy`) ve **2)** `config.example.yaml` dosyasındaki optimize edilmiş GEMMA hiperparametrelerini doğru bir şekilde okuyup kullanacak şekilde onarılacaktır.

#### 1. Özellik Maskesi Yükleme Hatasını Gider (`train_all_models.py`)

**Sorun:** Loglar, script'in `data/cache/gemma/feature_selection_mask.npy` dosyasını bulamadığını gösteriyor. Bu genellikle GitHub Actions ortamındaki göreceli yol (relative path) sorunlarından kaynaklanır.

**Çözüm:** Dosya yolunu, script'in çalıştığı yere göre daha sağlam bir şekilde tanımla.

- **Dosyayı Aç:** `scripts/train_all_models.py` dosyasını aç.
- **Yükleme Mantığını Onar:** `train_gemma_model` fonksiyonunun içindeki veya bu fonksiyonu besleyen `load_prepared_gemma_data` fonksiyonundaki maske yükleme mantığını bul.
- **Yolu Sağlamlaştır:** `Path("data/cache/gemma/feature_selection_mask.npy")` olan yolu, script'in kök dizinden itibaren çalıştığını varsayarak aynı şekilde bırakabiliriz, ancak bu dosyanın `git add` ile repository'ye eklendiğinden ve `checkout` adımında geldiğinden emin olmalıyız. Copilot Agent, bu dosyanın varlığını kontrol eden bir mantık eklemeli ve bulamazsa **eğitimi durdurmalıdır.**

  ```python
  # scripts/train_all_models.py -> train_gemma_model fonksiyonu içinde
  
  mask_path = Path('data/cache/gemma/feature_selection_mask.npy')
  if not mask_path.exists():
      logger.error(f"❌ KRİTİK HATA: Özellik seçim planı ({mask_path}) bulunamadı. Bu dosyanın repository'de olması gerekir. Eğitim durduruluyor.")
      # Fallback yapmak yerine süreci tamamen durdurmak, yanlış model üretmeyi engeller.
      raise FileNotFoundError(f"Feature selection mask not found at {mask_path}")

  logger.info(f"✅ Özellik seçim planı bulundu: {mask_path}")
  feature_mask = np.load(mask_path)
  X_selected = X_data_full[:, feature_mask]
  
  logger.info(f"✅ Özellik planı başarıyla uygulandı. {X_data_full.shape[1]} -> {X_selected.shape[1]} özellik.")
  ```

#### 2. Konfigürasyon Aktarma Hatasını Gider (`train_all_models.py`)

**Sorun:** Loglar, `RegimeModelTrainer`'ın konfigürasyon almadığını (`Config not provided`) ve bu yüzden fallback parametreleri kullandığını gösteriyor.

**Çözüm:** `main` fonksiyonunda yüklenen `ml_config` nesnesini, `train_gemma_model` fonksiyonuna ve oradan da `RegimeModelTrainer`'a doğru şekilde aktar.

- **Dosyayı Aç:** `scripts/train_all_models.py` dosyasını aç.
- **`train_gemma_model` Çağrısını Kontrol Et:** `main` fonksiyonunun en altındaki `train_gemma_model` çağrısının, `ml_config`'i parametre olarak geçtiğinden emin ol. (Mevcut kodda bu zaten doğru görünüyor).
- **`RegimeModelTrainer` Başlatmasını Düzelt:** `train_gemma_model` fonksiyonu içinde `RegimeModelTrainer` sınıfını başlatan satırı bul ve `config` parametresini doğru şekilde ilet.

  ```python
  # scripts/train_all_models.py -> train_gemma_model fonksiyonu içinde
  
  # Mevcut kodda, config nesnesi fonksiyona parametre olarak geliyor.
  # Bu config'in içinden sadece gemma ile ilgili bölümü alıp trainer'a vermeliyiz.
  
  # DOĞRU YAPI:
  gemma_config = config.get('gemma', {})
  if not gemma_config:
      raise ValueError("Konfigürasyonda 'gemma' bölümü bulunamadı.")
      
  # Trainer'ı bu spesifik konfigürasyon ile başlat.
  trainer = RegimeModelTrainer(config=gemma_config)
  
  # Modeli, seçilmiş ve ölçeklenmiş veri ile eğit.
  results = trainer.train_and_evaluate(X_selected, y_data_full, model_type=f'gemma_{model_type}')
  ```

#### 3. (YENİ GÖREV) Özellik Planı Doğrulamasını Ekle (train_all_models.py)
Sorun: Sistem, sadece .npy maskesine güveniyor ve .json dosyasını yok sayıyor. Çözüm: .json dosyasını yükleyip, seçilen özellik sayısının beklendiği gibi (82) olduğunu doğrula.

Dosyayı Aç: scripts/train_all_models.py

JSON Yükleme Mantığı Ekle: train_gemma_model fonksiyonuna, model_type'a göre doğru .json dosyasını yükleyen bir mantık ekle.

# train_gemma_model fonksiyonu içinde, maskeyi yükledikten sonra:

# Kontrol ve Denge Mekanizması:
json_plan_name = f"gemma_{model_type.split('_')[-1]}_selected_82.json"
json_plan_path = Path(f"features/gemma/selected/{json_plan_name}")

if not json_plan_path.exists():
    logger.error(f"❌ KRİTİK: Doğrulama için özellik listesi ({json_plan_path}) bulunamadı. Eğitim durduruluyor.")
    raise FileNotFoundError(f"Feature list JSON not found at {json_plan_path}")
    
with open(json_plan_path, 'r') as f:
    feature_plan = json.load(f)

selected_feature_count_from_json = feature_plan.get('count', 0)
selected_feature_count_from_mask = np.sum(feature_mask)

if selected_feature_count_from_json != selected_feature_count_from_mask:
    logger.error(f"❌ KRİTİK: Maske ve JSON planı arasında tutarsızlık! Maske: {selected_feature_count_from_mask}, JSON: {selected_feature_count_from_json}. Eğitim durduruluyor.")
    raise ValueError("Feature mask and JSON plan are inconsistent.")
    
logger.info(f"✅ Özellik planı doğrulandı: {json_plan_path} (Beklenen: {selected_feature_count_from_json} özellik)")

#### 4. MLP Mimarisi Çevirim Hatasını Gider (`model_trainer.py`)

**Sorun:** Konfigürasyon alınamadığı için, `RegimeModelTrainer` hatalı bir "LSTM'den MLP'ye çevrim" mantığı kullanıyor.

**Çözüm:** `RegimeModelTrainer`'ın, `self.config` üzerinden gelen doğru `gemma.architecture` bloğunu okumasını sağla.

- **Dosyayı Aç:** `src/ml/model_trainer.py` dosyasını aç.
- **`train_and_evaluate` Metodunu Güncelle:** Bu metodun içindeki, MLP modelini oluşturan bölümü bul.
- **Doğru Parametreleri Kullan:** Bu bölümün, `self.config['architecture']` içindeki `hidden_size`, `num_layers`, `input_size` gibi değerleri doğrudan okuduğundan emin ol. Hatalı çevrim mantığını kaldır.

  ```python
  # src/ml/model_trainer.py -> train_and_evaluate içinde

  # 'gemma' modeli için MLP oluşturma bloğu:
  if 'gemma' in model_type:
      logger.info(f"Training MLP model for {model_type}")
      
      # Artık config'in doğru geldiğini varsayıyoruz.
      arch_config = self.config.get('architecture', {})
      
      # Parametreleri doğrudan config'den oku
      input_size = X.shape[1] # Gerçek veri setinin özellik sayısını kullan!
      hidden_size = arch_config.get('hidden_size')
      num_layers = arch_config.get('num_layers')
      dropout = arch_config.get('dropout')
      num_classes = arch_config.get('num_classes')

      if not all([hidden_size, num_layers, dropout, num_classes]):
          raise ValueError("GEMMA mimari parametreleri (hidden_size, num_layers, dropout, num_classes) config'de eksik.")

      # MLP katmanlarını doğru oluştur
      # Örneğin: layers = [hidden_size] * num_layers -> [64, 64] gibi
      layers = [hidden_size for _ in range(num_layers)]

      logger.info("MLP Configuration from config.example.yaml:")
      logger.info(f"  Hidden Layers: {layers}")
      # ... diğer loglar ...

      # Modeli bu doğru parametrelerle oluştur.
      model = GemmaMLP(input_size=input_size, layers=layers, ...)
      
      # ... (kalan eğitim kodları) ...
  ```

### ✅ Başarı Kriterleri

- `train-models.yml` workflow'u çalıştırıldığında, loglarda artık **hiçbir `WARNING` veya `ERROR` mesajı görülmemelidir.**
- Loglarda, **"Özellik seçim planı başarıyla uygulandı. 87 -> 82 özellik."** gibi bir mesaj açıkça yer almalıdır.
- (YENİ) Loglarda, "Özellik planı doğrulandı: features/gemma/selected/gemma_price_selected_82.json (Beklenen: 82 özellik)" gibi bir mesaj görülmelidir.
- Loglarda, **"MLP Configuration from config.example.yaml"** başlığı altında, `config.example.yaml` dosyasındaki `gemma.architecture` bölümüyle eşleşen hiperparametrelerin listelendiği görülmelidir.
- Eğitim sonunda ulaşılan test doğruluk (`Accuracy`) oranının, hedeflenen ~%78 seviyesine önemli ölçüde yaklaşması gerekir.

Bu görev tamamlandığında, `train-models.yml` workflow'u, projenin en başından beri hedeflenen, mimariye tam uyumlu, "doğru çalışan" bir üretim hattı haline gelecektir.

**Comment by SefaGH** [comment]
#### **Critical Action: Environment Setup and Verification**

1.  **Enforce Python Version 3.11:**
    *   The project requires Python `3.11.x`. The runner environment may default to a different version (e.g., 3.12).
    *   **Action:** Your first step must be to check the active Python version (`python --version`). If it is not `3.11`, you must stop, install Python 3.11, and switch the environment to use it exclusively.

2.  **Mandatory Full Dependency Installation:**
    *   The validation requires all project dependencies, including heavy libraries like `torch`, `pandas`, and `scikit-learn`. Skipping any package will lead to invalid test results.
    *   **Action:** After ensuring Python 3.11 is active, perform a complete and non-simplified installation of all packages listed in the project's dependency files (e.g., `requirements.txt`). Do not use mocked libraries for this validation.

---

#### **Execution Directive**

Once the environment is confirmed to be running **Python 3.11** and **all dependencies are fully installed**, proceed with executing the complete validation plan as detailed in the issue description above, starting from `Task 1`.

The goal is a full, non-simplified validation of the artifacts and code.
